feat(Vizzy): new presence

BigBoi-TM opened the pull request:
## Description 
<!-- A clear and detailed description of the changes, referencing issues if applicable -->
Resolves #6030
Features:

- Detects homepage, community page, user page, editor, export page, etc.
   - Editor: detects when editing owned projects or community projects
   - Exporter: displays exporting status (I.e.: waiting, estimated time left, etc.)

`*Note: most elements are found using xPath because they have random id's`
## Acknowledgements
- [x] I read the [Presence Guidelines](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md)
- [x] I linted the code by running `yarn format`

## Screenshots
<details>
<summary> Proof showing the creation/modification is working as expected </summary>
<!-- 
    Screenshots of the presence settings (if applicable) and at least TWO screenshots of the presence displaying correctly
    Including these screenshots will assist the reviewing processes thus speeding up the process of the pull request being merged
-->

### - Homepage
![image](https://user-images.githubusercontent.com/96930330/169951030-f2c40508-0860-4d9e-9116-2aca56eefaa7.png)

### - Community
`P.S.: Same page goes to two paths`
![image](https://user-images.githubusercontent.com/96930330/169951264-5145c16a-efac-46c5-af53-cab406b79ad5.png)
![image](https://user-images.githubusercontent.com/96930330/169951387-a78fe009-baab-4ac2-ba94-47031fe38729.png)

### - Project
![image](https://user-images.githubusercontent.com/96930330/169951535-fb2c9e24-953b-412f-9ce7-6993a439c5ff.png)

### - User
![image](https://user-images.githubusercontent.com/96930330/169951638-9a32036a-ca98-4c9f-81e1-27e3e72edd87.png)

### - Editor
`v Community Project v`
![image](https://user-images.githubusercontent.com/96930330/169951900-93796b6e-118a-407a-8428-394e845c44f8.png)
`v Owned Project v`
![image](https://user-images.githubusercontent.com/96930330/169952049-3a75e6e8-5a3f-4f49-987c-f2fd96985083.png)

### - Exporting
`v Awaiting Export v`
![image](https://user-images.githubusercontent.com/96930330/169952212-bb3678fa-7690-4a66-94d9-1cb72850f1fd.png)
`v Exporting v`
![image](https://user-images.githubusercontent.com/96930330/169952249-aede97aa-666e-418c-999f-8b3e01982684.png)
![image](https://user-images.githubusercontent.com/96930330/169952293-db423d92-da01-4646-bbf7-6d90b14faae4.png)


</details>

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,122 @@
+			startTimestamp: browsingTimestamp
+		},
+		{ pathname, search } = window.location;
+	switch (true) {
```
all you do in this switch statement is compare a variable to a string so please do a proper switch statement with the variable in here and the cases down below

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,122 @@
+		case pathname === "/project": {
+			presenceData.largeImageKey = "discover";
+			presenceData.details = "Viewing a project";
+			presenceData.state = `Project: ${
+				getElementByXpath(
+					"/html/body/div[1]/div/main/div/div/div[2]/div/div[2]/div[1]/h4"
+				).textContent
+			}`;
```
```suggestion
			presenceData.state =
				getElementByXpath(
					"/html/body/div[1]/div/main/div/div/div[2]/div/div[2]/div[1]/h4"
				).textContent;
```

BigBoi-TM replied:
Can't keep the `Project:` part?

ImRodry replied:
No, you already mention the user is viewing a project in the presence details right above

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,122 @@
+		case pathname === "/profile": {
+			presenceData.details = "Viewing their profile";
+			presenceData.largeImageKey = "profile";
+			if (search) presenceData.state = `Tab: ${search.split("=")[1]}`;
```
this isn't reliable, you should use URLSearchParams or a regex

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,122 @@
+				).textContent;
+			} else {
+				presenceData.details = "Preparing to export a video";
+				/* if (document.querySelector<HTMLDivElement>('id="notistack-snackbar'))
+					presenceData.state = "Status: Export Complete..."; */
```
remove

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,122 @@
+				presenceData.details = "Preparing to export a video";
+				/* if (document.querySelector<HTMLDivElement>('id="notistack-snackbar'))
+					presenceData.state = "Status: Export Complete..."; */
+				presenceData.state = "Status: Waiting...";
```
do you really need this here?

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,122 @@
+			break;
+		}
+		case pathname === "/discover": {
+			if (search === "?tab=faq") {
```
you should use URLSearchParams

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,122 @@
+			}
+			break;
+		}
+		case pathname === "/privacy": {
+			presenceData.largeImageKey = "faq";
```
you sure you want to put the faq image key on the privacy page?

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,129 @@
+const presence = new Presence({
+		clientId: "968353669491871754"
+	}),
+	browsingTimestamp = Math.floor(Date.now());
```
what are you flooring? An integer?

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,129 @@
+			presenceData.buttons = [
+				{
+					label: "View Project",
+					url: location.href
```
Declare href in the destructure at line 19
```suggestion
					url: href
```

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,129 @@
+			presenceData.buttons = [
+				{
+					label: "View User",
+					url: location.href
```
```suggestion
					url: href
```

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,129 @@
+			smallImageKey: "logo_512",
+			startTimestamp: browsingTimestamp
+		},
+		{ pathname, search } = window.location;
```
```suggestion
		{ pathname, search, href } = window.location;
```

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,129 @@
+			presenceData.largeImageKey = "profile";
+			if (search) {
+				presenceData.state = `Tab: ${new URLSearchParams(
+					document.location.search
```
make some sense please
```suggestion
					search
```

BigBoi-TM replied:
Cuz it was addressed; replaced with the search variable `tabParams`

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,129 @@
+			break;
+		}
+		case "/discover": {
+			if (new URLSearchParams(document.location.search).get("tab") === "faq") {
```
just declare the searchparams variable at the top

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,129 @@
+		}
+		case "/privacy": {
+			presenceData.largeImageKey = "discover";
+			presenceData.details = "Reading privacy terms";
```
```suggestion
			presenceData.details = "Reading the privacy policy";
```

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,129 @@
+		}
+		case "/terms-of-service": {
+			presenceData.largeImageKey = "discover";
+			presenceData.details = "Reading terms of use";
```
```suggestion
			presenceData.details = "Reading the terms of use";
```

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,127 @@
+		}
+		case "/export": {
+			presenceData.largeImageKey = "export";
+			if (getElementByXpath("/html/body/div[1]/div/div/div[2]/div[2]/p")) {
+				presenceData.details = "Exporting a video";
+				presenceData.state = getElementByXpath(
+					"/html/body/div[1]/div/div/div[2]/div[2]/p"
+				).textContent;
```
should probably only compute this once

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,127 @@
+			break;
+		}
+		case "/discover": {
+			const tabParams = new URLSearchParams(search).get("tab");
```
I believe I suggested making a searchParams variable at the top so you don't have to compute this so often

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,119 @@
+		case "/profile": {
+			presenceData.details = "Viewing their profile";
+			presenceData.largeImageKey = "profile";
+			if (search) presenceData.state = `Tab: ${tabParams}`;
```
```suggestion
			if (tabParams) presenceData.state = `Tab: ${tabParams}`;
```

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,119 @@
+			break;
+		}
+		case "/discover": {
+			// eslint-disable-next-line unicorn/prefer-switch
```
uhm no? Don't disable it

ImRodry committed:
feat(Vizzy): new presence

feat(Vizzy): new presence

feat(Vizzy): new presence

style: fix formatting issues

refactor: switched `imageKey`'s and removed `eslint-disable` line
Co-Authored-By: Slowlife <slowlife1165@gmail.com>
Co-Authored-By: Rodry <38259440+ImRodry@users.noreply.github.com>

style: fixes styling issues

Update presence.ts

ImRodry commented:
Rebased to apply the latest changes in the prettier configuration. I also squashed all the commits in your branch as the commit history was quite messy. If you need help with your local branch, please ping me on the #presence-dev channel on Discord.

BigBoi-TM commented:
bruh

BigBoi-TM commented:
finally 😤

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,127 @@
+					break;
+				}
+				default:
+					if (!search) presenceData.details = "Browsing features";
```
what if there is search and there isn't a tab param in search? Wouldn't you want to display the same thing?

ImRodry committed:
Merge branch 'main' into vizzy-1.0.0

ImRodry commented on websites/V/Vizzy/presence.ts:
```diff
@@ -0,0 +1,128 @@
+					break;
+				}
+				default:
+					if ((search && !tabParams) || !search)
```
Works the same
```suggestion
					if (!tabParams || !search)
```

ImRodry committed:
Update websites/V/Vizzy/presence.ts

ImRodry commented:
what even happened here

BigBoi-TM commented:
Don't look at me, I ain't touch nothing

ImRodry committed:
style: fix styling issues

ImRodry commented:
there are no prettier issues, the CI is definitely not ok, we'll have to fix this

ImRodry committed:
Merge branch 'main' into vizzy-1.0.0
